vnext/ReactWindowsCore/WinRTWebSocketResource.cpp

```diff
@@ -0,0 +1,434 @@
+using winrt::Windows::Foundation::Uri;
+using winrt::Windows::Networking::Sockets::IWebSocket;
+using winrt::Windows::Networking::Sockets::MessageWebSocket;
+using winrt::Windows::Networking::Sockets::MessageWebSocketMessageReceivedEventArgs;
+using winrt::Windows::Networking::Sockets::SocketMessageType;
+using winrt::Windows::Networking::Sockets::WebSocketClosedEventArgs;
+using winrt::Windows::Security::Cryptography::Certificates::ChainValidationResult;
+using winrt::Windows::Security::Cryptography::CryptographicBuffer;
+using winrt::Windows::Storage::Streams::DataReader;
+using winrt::Windows::Storage::Streams::DataWriter;
+using winrt::Windows::Storage::Streams::UnicodeEncoding;
+
+namespace
+{
+  ///
+  /// Implements an awaiter for Mso::DispatchQueue
+  ///
+  auto resume_in_queue(const Mso::DispatchQueue& queue)
+  {
+    struct awaitable
+    {
+      awaitable(const Mso::DispatchQueue& queue)
+        : m_queue{ queue }
+      {
+      }
+
+      bool await_ready() const noexcept
+      {
+        return false;
+      }
+
+      void await_resume() const noexcept {}
+
+      void await_suspend(std::experimental::coroutine_handle<> resume)
+      {
+        m_queue.Post([context = resume.address()]() noexcept
+        {
+          std::experimental::coroutine_handle<>::from_address(context)();
+        });
+      }
+
+    private:
+      Mso::DispatchQueue m_queue;
+    };
+
+    return awaitable{ queue };
+  }
+}
+
+namespace Microsoft::React
+{
+WinRTWebSocketResource::WinRTWebSocketResource(Uri&& uri, vector<ChainValidationResult> certExeptions)
+  : m_uri{ std::move(uri) }
+{
+  m_socket.MessageReceived({ this, &WinRTWebSocketResource::OnMessageReceived });
+
+  for (auto certException : certExeptions)
+  {
+    m_socket.Control().IgnorableServerCertificateErrors().Append(certException);
+  }
+}
+
+WinRTWebSocketResource::WinRTWebSocketResource(const string& urlString, vector<ChainValidationResult> certExeptions)
+  : WinRTWebSocketResource(Uri{ Utf8ToUtf16(urlString) }, certExeptions)
+{
+}
+
+WinRTWebSocketResource::~WinRTWebSocketResource() /*override*/
+{
+  Close(CloseCode::GoingAway, "Disposed");
+
+  m_closePerformed.Wait();
+}
+
+#pragma region Private members
+
+IAsyncAction WinRTWebSocketResource::PerformConnect()
+{
+  auto self = shared_from_this();
+  try
+  {
+    co_await resume_background();
+
+    co_await self->m_socket.ConnectAsync(self->m_uri);
+
+    self->m_readyState = ReadyState::Open;
+    if (self->m_connectHandler)
+    {
+      self->m_connectHandler();
+    }
+  }
+  catch (hresult_error const& e)
+  {
+    self->m_readyState = ReadyState::Closed;
+    if (self->m_errorHandler)
+    {
+      self->m_errorHandler({ Utf16ToUtf8(e.message()), ErrorType::Connection });
+    }
+  }
+
+  SetEvent(self->m_connectPerformed.get());
+  self->m_connectPerformedPromise.set_value();
+  self->m_connectRequested = false;
+}
+
+fire_and_forget WinRTWebSocketResource::PerformPing()
```

Comment:
We might want to use the dispatchqueue here to. E.g. to preserve ordering between pings/writes.

Reply:
Writes and pings should be handled independently, per standard.
They are not expected to have a relative order of execution.